<a href="https://colab.research.google.com/github/1Piyusharma/ML/blob/main/linear%20regression%20ml%20project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data Handling

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('/content/student_placement_prediction_dataset_2026.csv')

df.head()

,student_id,age,gender,cgpa,branch,college_tier,internships_count,projects_count,certifications_count,coding_skill_score,...,mock_interview_score,attendance_percentage,backlogs,extracurricular_score,leadership_score,volunteer_experience,sleep_hours,study_hours_per_day,placement_status,salary_package_lpa
0,1,24,Male,7.53,IT,Tier 2,4,6,1,99.238568,...,72.647009,77.463863,2,63.382726,52.938240,Yes,6.7,3.6,Not Placed,0.00
1,2,21,Male,7.92,CSE,Tier 2,1,3,6,80.966123,...,61.699110,88.887600,1,73.694605,60.198856,No,4.4,2.3,Not Placed,0.00
2,3,22,Female,8.60,EEE,Tier 1,0,1,1,49.177184,...,87.396911,74.153265,0,63.329294,43.708803,No,8.8,5.9,Placed,11.99
3,4,24,Male,6.68,CSE,Tier 1,0,2,2,79.359084,...,58.401069,87.635955,1,47.636099,56.549154,Yes,8.1,4.4,Not Placed,0.00
4,5,20,Female,8.43,IT,Tier 3,1,4,3,65.018573,...,74.489201,79.120749,1,0.000000,67.268893,No,8.7,3.4,Placed,12.16


In [3]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 26 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   student_id                 100000 non-null  int64  
 1   age                        100000 non-null  int64  
 2   gender                     100000 non-null  object 
 3   cgpa                       100000 non-null  float64
 4   branch                     100000 non-null  object 
 5   college_tier               100000 non-null  object 
 6   internships_count          100000 non-null  int64  
 7   projects_count             100000 non-null  int64  
 8   certifications_count       100000 non-null  int64  
 9   coding_skill_score         100000 non-null  float64
 10  aptitude_score             100000 non-null  float64
 11  communication_skill_score  100000 non-null  float64
 12  logical_reasoning_score    100000 non-null  float64
 13  hackathons_participated    100

,student_id,age,cgpa,internships_count,projects_count,certifications_count,coding_skill_score,aptitude_score,communication_skill_score,logical_reasoning_score,...,github_repos,linkedin_connections,mock_interview_score,attendance_percentage,backlogs,extracurricular_score,leadership_score,sleep_hours,study_hours_per_day,salary_package_lpa
count,100000.000000,100000.000000,100000.000000,100000.00000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000,...,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000
mean,50000.500000,21.005740,7.496890,1.50223,3.001860,2.005150,69.825326,64.988108,67.960744,65.931696,...,4.005160,524.559110,69.989411,84.730169,0.497590,59.894049,54.937526,6.983319,3.523124,7.252106
std,28867.657797,1.999562,0.992834,1.22528,1.731764,1.416553,14.694618,14.844093,13.846636,14.826790,...,2.004897,274.993479,11.937257,9.416075,0.708434,19.549666,19.781197,1.468113,1.471248,6.735132
min,1.000000,18.000000,4.500000,0.00000,0.000000,0.000000,20.000000,20.000000,20.000000,20.000000,...,0.000000,50.000000,20.000000,50.000000,0.000000,0.000000,0.000000,3.000000,0.500000,0.000000
25%,25000.750000,19.000000,6.820000,1.00000,2.000000,1.000000,59.807945,54.894600,58.587535,55.940547,...,3.000000,287.000000,61.973172,78.325833,0.000000,46.538590,41.432390,6.000000,2.500000,0.000000
50%,50000.500000,21.000000,7.500000,1.00000,3.000000,2.000000,70.006250,65.011394,67.973221,66.003829,...,4.000000,523.000000,70.026665,85.018828,0.000000,60.079021,55.076210,7.000000,3.500000,11.100000
75%,75000.250000,23.000000,8.180000,2.00000,4.000000,3.000000,80.058762,75.079266,77.503749,76.113705,...,5.000000,764.000000,78.122083,91.782511,1.000000,73.549567,68.530115,8.000000,4.500000,13.470000
max,100000.000000,24.000000,10.000000,8.00000,13.000000,11.000000,100.000000,100.000000,100.000000,100.000000,...,16.000000,999.000000,100.000000,100.000000,6.000000,100.000000,100.000000,10.000000,10.000000,20.440000


In [4]:
df.isnull().sum()

# Drop null values
df = df.dropna()

In [5]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for col in df.select_dtypes(include=['object']).columns:
    df[col] = le.fit_transform(df[col])

In [6]:
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1

df = df[~((df < (Q1 - 1.5 * IQR)) | (df > (Q3 + 1.5 * IQR))).any(axis=1)]

In [7]:
X = df.drop('salary_package_lpa', axis=1)
y = df['salary_package_lpa']

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

Model Training & Hyperparameter Tuning

In [9]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Baseline Results:")
print("MSE:", mse)
print("R2 Score:", r2)

Baseline Results:
MSE: 0.8640542364414951
R2 Score: 0.9806453283146169


In [10]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model_scaled = LinearRegression()
model_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = model_scaled.predict(X_test_scaled)

mse_scaled = mean_squared_error(y_test, y_pred_scaled)
r2_scaled = r2_score(y_test, y_pred_scaled)

print("After Normalization:")
print("MSE:", mse_scaled)
print("R2 Score:", r2_scaled)

After Normalization:
MSE: 0.8640542364414951
R2 Score: 0.9806453283146169


In [11]:
print("Before Scaling R2:", r2)
print("After Scaling R2:", r2_scaled)

Before Scaling R2: 0.9806453283146169
After Scaling R2: 0.9806453283146169


Prediction


In [13]:
import pandas as pd

# Get the exact column names from the X DataFrame used for training
# X_train is a numpy array after scaling, so we get column names from original X
x_columns = X.columns.tolist()

# Sample students (ensure column names and order EXACTLY match your dataset 'X')
# Also ensure categorical features are encoded as done during preprocessing
# For demonstration, I will provide example values for two students.
# The values for 'gender', 'branch', 'college_tier', 'volunteer_experience', 'placement_status'
# should reflect the label encoding performed earlier.
# Assuming:
#   - gender: Male=1, Female=0
#   - college_tier: Tier 1=0, Tier 2=1, Tier 3=2 (based on describe output min/max)
#   - volunteer_experience: No=0, Yes=1 (based on describe output min/max)
#   - placement_status: Not Placed=0, Placed=1
#   - branch: (encoding values would depend on the order the LabelEncoder encountered them, using example values here)

sample_data_values = [
    { # Sample Student 1
        'student_id': 100001,
        'age': 22,
        'gender': 1, # Male
        'cgpa': 8.5,
        'branch': 0, # Example encoded value for a branch (e.g., CSE)
        'college_tier': 0, # Tier 1
        'internships_count': 3,
        'projects_count': 5,
        'certifications_count': 2,
        'coding_skill_score': 90.0,
        'aptitude_score': 85.0,
        'communication_skill_score': 80.0,
        'logical_reasoning_score': 82.0,
        'hackathons_participated': 3,
        'github_repos': 7,
        'linkedin_connections': 800,
        'mock_interview_score': 78.0,
        'attendance_percentage': 92.0,
        'backlogs': 0,
        'extracurricular_score': 70.0,
        'leadership_score': 75.0,
        'volunteer_experience': 1, # Yes
        'sleep_hours': 7.5,
        'study_hours_per_day': 4.0,
        'placement_status': 1 # Placed
    },
    { # Sample Student 2
        'student_id': 100002,
        'age': 20,
        'gender': 0, # Female
        'cgpa': 6.8,
        'branch': 1, # Example encoded value for another branch (e.g., IT)
        'college_tier': 2, # Tier 3
        'internships_count': 1,
        'projects_count': 2,
        'certifications_count': 1,
        'coding_skill_score': 60.0,
        'aptitude_score': 55.0,
        'communication_skill_score': 65.0,
        'logical_reasoning_score': 58.0,
        'hackathons_participated': 0,
        'github_repos': 2,
        'linkedin_connections': 300,
        'mock_interview_score': 50.0,
        'attendance_percentage': 80.0,
        'backlogs': 1,
        'extracurricular_score': 45.0,
        'leadership_score': 30.0,
        'volunteer_experience': 0, # No
        'sleep_hours': 6.0,
        'study_hours_per_day': 2.5,
        'placement_status': 0 # Not Placed
    }
]

sample_data = pd.DataFrame(sample_data_values, columns=x_columns)

# If you also want to apply scaling to the sample data before prediction, use the same scaler
# X_test_scaled was generated using 'scaler.transform(X_test)', so we use scaler.transform here.
# For predicting with the 'model' (which was trained on unscaled data), we use 'sample_data' directly.
# If predicting with 'model_scaled', you would need:
# sample_data_scaled = scaler.transform(sample_data)
# predictions = model_scaled.predict(sample_data_scaled)

# Predict using the unscaled model
predictions = model.predict(sample_data)

print(predictions)

[14.58186207 -0.60875972]


 This is salary prediction (LPA)

* 14.58 → very high package (strong candidate)
* -0.60 → mathematically possible, but real-world invalid (means “very low → effectively not placed”)

In [14]:
# Convert salary → placement
placement_pred = (predictions > 0).astype(int)

print("Raw Predictions (Salary):", predictions)
print("Placement Prediction:", placement_pred)

Raw Predictions (Salary): [14.58186207 -0.60875972]
Placement Prediction: [1 0]


In [15]:
predictions = model.predict(sample_data)

placement_pred = (predictions > 0).astype(int)

print(placement_pred)

[1 0]
